libbuf/libbuf.js

```diff
@@ -0,0 +1,224 @@
+// This class exists just to pick the correct operation class,
+// when receiving operations (JSON) from remote clients/server
+// it is a class and not a function for consistency, the API matches
+// the other classes; Operation.from({}), Insert.from({}) etc.
+class Operation {
+  static from(data) {
+    switch (data.command) {
+      case "content":
+        return Content.from(data);
+      case "insert":
+        return Insert.from(data);
+      case "remove":
+        return Remove.from(data);
+      default:
+        throw "Unkown operation: " + data.command;
+    }
+  }
+
+  copy() {
+    return Operation.from(this);
+  }
+
+  identical(other) {
+    return (
+      this.time === other.time &&
+      this.command === other.command &&
+      this.row === other.row &&
+      this.column === other.column &&
+      this.string === other.string
+    );
+  }
+
+  before(other) {
+    return new Date(this.time) < new Date(other.time);
+  }
+
+  after(other) {
+    return new Date(this.time) > new Date(other.time);
+  }
+
+  json() {
+    return JSON.stringify(this);
+  }
+}
+
+class Insert extends Operation {
+  constructor(string, row, column, time = new Date().toISOString()) {
+    super();
+    this.time = time;
+    this.command = "insert";
+    this.string = string;
+    this.row = row;
+    this.column = column;
+  }
+
+  static from(data) {
+    return new Insert(data.string, data.row, data.column, data.time);
+  }
+
+  apply(content) {
+    let lines = this.string.split("\n");
+    let row = this.row;
+    let column = this.column;
+
+    let before = content.slice(0, row);
+    let affected = content[row];
+    let extra = lines.slice(1);
+    let after = content.slice(row + 1);
+
+    let edited = affected.slice(0, column) + lines[0] + affected.slice(column);
+
+    let result = before.concat(edited, extra, after);
+    return result;
+  }
+}
+
+class Remove extends Operation {
+  constructor(string, row, column, time = new Date().toISOString()) {
+    super();
+    this.string = string;
+    this.row = row;
+    this.column = column;
+    this.time = time;
+    this.command = "remove";
+  }
+
+  static from(data) {
+    return new Remove(data.string, data.row, data.column, data.time);
+  }
+
+  apply(content) {
+    let to_remove = this.string.split("\n");
+    let number_of_lines = to_remove.length - 1;
+    let row = this.row;
+    let column = this.column;
+
+    let before = content.slice(0, row);
+
+    let affected = content.slice(row, row + to_remove.length);
+    let msg = JSON.stringify(affected) + " -> ";
+    if (affected.length > 2) {
+      affected = [affected[0], affected[affected.length - 1]];
+      to_remove = [to_remove[0], to_remove[to_remove.length - 1]];
+      msg = msg + JSON.stringify(affected) + " -> ";
+    }
+    affected[0] =

```

Comment:
Tror du kan bruke splice istedenfor slice her. Kanskje andre steder også 😄 
Bytt ut linje 106 med det du vil fjerne i to og from. 

affected[0].splice(to, from)

Reply:
Kan ikke splice strings, de er immutable. (Måtte i så fall konvertert til array, splice, konvertert tilbake, det er ikke noe penere, IMO).